crates/rethnet_evm/src/miner.rs

```diff
@@ -43,155 +42,108 @@ pub enum MineBlockError<BE, SE> {
+                    Some(prevrandao.ok_or(MineBlockError::MissingPrevrandao)?)
+                } else {
+                    None
                 },
-            )
-            .await?
-        };
-
-        let mut transaction_pool = self.mem_pool.write().await;
-        let mut pending_transactions: VecDeque<_> = transaction_pool
-            .pending_transactions()
-            .iter()
-            .cloned()
-            .collect();
-
-        let mut results = Vec::new();
-        let mut traces = Vec::new();
-
-        while let Some(transaction) = pending_transactions.pop_front() {
-            let mut tracer = TraceCollector::default();
-
-            let transaction_hash = *transaction.hash();
-
-            match block_builder
-                .add_transaction(transaction, Some(&mut tracer))
-                .await
-            {
-                Err(BlockTransactionError::ExceedsBlockGasLimit) => continue,
-                Err(e) => {
-                    block_builder
-                        .abort()
-                        .await
-                        .map_err(MineBlockError::BlockAbort)?;
-
-                    return Err(MineBlockError::BlockTransaction(e));
-                }
-                Ok(result) => {
-                    results.push(result);
-                    traces.push(tracer.into_trace());
-
-                    transaction_pool.remove_transaction(&transaction_hash);
-                }
+                nonce: Some(if cfg.spec_id >= SpecId::MERGE {
+                    B64::ZERO
+                } else {
+                    B64::from(U64::from(42))
+                }),
+                base_fee: if cfg.spec_id >= SpecId::LONDON {
+                    Some(base_fee.unwrap_or_else(|| calculate_next_base_fee(&parent_block.header)))
+                } else {
+                    None
+                },
+                ..Default::default()
+            },
+        )?
+    };
+
+    let mut pending_transactions: VecDeque<_> =
+        mem_pool.pending_transactions().iter().cloned().collect();
+
+    let mut results = Vec::new();
+    let mut traces = Vec::new();
+
+    while let Some(transaction) = pending_transactions.pop_front() {
+        let mut tracer = TraceCollector::default();
+
+        let transaction_hash = *transaction.hash();
+
+        match block_builder.add_transaction(blockchain, state, transaction, Some(&mut tracer)) {
+            Err(BlockTransactionError::ExceedsBlockGasLimit) => continue,
+            Err(e) => {
+                block_builder
+                    .abort(state)
+                    .map_err(MineBlockError::BlockAbort)?;
+
+                return Err(MineBlockError::BlockTransaction(e));
+            }
+            Ok(result) => {
+                results.push(result);
+                traces.push(tracer.into_trace());
+
+                mem_pool.remove_transaction(&transaction_hash);
             }
         }
+    }
```

Comment:
Good find!

I have several other commits with fixes for the block builder, miner, and other types waiting for this to be merged; to avoid too many conflicts. Do you mind if I address this there, as this doesn't actually change the logic?

I created a ticket for this: https://github.com/NomicFoundation/rethnet/issues/159

Reply:
Don't mind at all, didn't mean to suggest you should address it here, thanks for tracking it.